packages/internal-test-env/index.ts

```diff
@@ -0,0 +1,357 @@
+//
+// Copyright 2022 Inrupt Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+// Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+import { config } from "dotenv-flow";
+import { join } from "path";
+import { Session } from "@inrupt/solid-client-authn-node";
+import {
+  createContainerInContainer,
+  createSolidDataset,
+  deleteSolidDataset,
+  getPodUrlAll,
+  getSourceIri,
+  saveSolidDatasetInContainer,
+} from "@inrupt/solid-client";
+
+const availableEnvironment = [
+  "ESS Dev-Next" as const,
+  "ESS PodSpaces" as const,
+  "NSS" as const,
+  //  "ESS PodSpaces Next" as const,
+];
+
+export type AvailableEnvironment = typeof availableEnvironment extends Array<
+  infer E
+>
+  ? E
+  : never;
+
+const availableProtocol = ["ESS Notifications Protocol" as const];
+
+export type AvailableProtocol = typeof availableProtocol extends Array<infer E>
+  ? E
+  : never;
+
+export interface TestingEnvironmentNode {
+  environment: AvailableEnvironment;
+  idp: string;
+  notificationGateway: string;
```

Comment:
The `notificationGateway` is only used in the notifications test: in other repositories, it will be undefined. However, it being present is enforced by `getTestingEnvironment`.

Should more flexibility be given here by maybe having a `librarySpecific` section of the environment, which could be a `Record<string, string>`, with an env var pattern used to match them at instantiation, pretty much like it is done in the feature flags? 

The env could have something like `E2E_TEST_LIB_...`, e.g. `E2E_TEST_LIB_NOTIFICATION_GATEWAY`, which would result in the `TestingEnvironment` to contain
```
{
  environment: "...",
  librarySpecific: {
    NOTIFICATION_GATEWAY: "..."
  }
}
```
The library-specific environment could extend the shared one to make this typing explicit.

I don't want to derail the scope of this ticket: we can also put a pin on this, just make sure that the environments will validate across all repos by taking the smallest common denominator, and we can extend on this later.

Reply:
tbh, I don't think `notificationGateway` is even needed — it should be doing storage metadata discovery.. I think this is a left over maybe?

Reply:
The other alternative is a generic, so: 
```typescript
getNodeTestingEnvironment<{ notificationGateway: string }>((env) => {
  if (typeof env.E2E_TEST_NOTIFICATION_GATEWAY === "string") {
    return { notificationGateway: env.E2E_TEST_NOTIFICATION_GATEWAY };
  } else {
    throw new Error("Missing E2E_TEST_NOTIFICATION_GATEWAY environment variable");
  }
})
```

Or something

Reply:
Okay Ill see if I can remove notificationsGateway if not I will make it optional